src/Nixfmt/Types.hs

```diff
@@ -0,0 +1,242 @@
+
+data Binder = Inherit Leaf (Maybe Term) [Leaf] Leaf
+            | Assignment [Selector] Leaf Expression Leaf
+            | BinderTrivia Trivia
+            deriving (Show)
+
+data ListPart = ListItem Term
+              | ListTrivia Trivia
+              deriving (Show)
+
+data Term = Token Leaf
+          | String String
+          | List Leaf [ListPart] Leaf
+          | Set (Maybe Leaf) Leaf [Binder] Leaf
+          | Selection Term [Selector]
+          | Parenthesized Leaf Expression Leaf
+          deriving (Show)
+
+data ParamAttr = ParamAttr Leaf (Maybe (Leaf, Expression)) (Maybe Leaf)
+               | ParamEllipsis Leaf
+               deriving (Show)
+
+data Parameter = IDParameter Leaf
+               | SetParameter Leaf [ParamAttr] Leaf
+               | ContextParameter Parameter Leaf Parameter
+               deriving (Show)
+
+data Expression = Term Term
+                | With Leaf Expression Leaf Expression
+                | Let Leaf [Binder] Leaf Expression
+                | Assert Leaf Expression Leaf Expression
+                | If Leaf Expression Leaf Expression Leaf Expression
+                | Abstraction Parameter Leaf Expression
+
+                | Application Expression Expression
+                | Operation Expression Leaf Expression
+                | MemberCheck Expression Leaf Selector
+                | Negation Leaf Expression
+                | Inversion Leaf Expression
+                deriving (Show)
+
+data File = File Leaf Expression
+          deriving (Show)
+
+data Token = Integer    Int
+           | Identifier Text
+           | Path       Text
+           | EnvPath    Text
+
+           | KAssert
+           | KElse
+           | KIf
+           | KIn
+           | KInherit
+           | KLet
+           | KOr
+           | KRec
+           | KThen
+           | KWith
+
+           | TBraceOpen
+           | TBraceClose
+           | TBrackOpen
+           | TBrackClose
+           | TInterOpen
+           | TInterClose
+           | TParenOpen
+           | TParenClose
+
+           | TAssign
+           | TAt
+           | TColon
+           | TComma
+           | TDot
+           | TDoubleQuote
+           | TDoubleSingleQuote
+           | TEllipsis
+           | TQuestion
+           | TSemicolon
+
+           | TConcat
+           | TNegate
+           | TUpdate
+
+           | TPlus
+           | TMinus
+           | TMul
+           | TDiv
+
+           | TAnd
+           | TOr
+           | TEqual
+           | TGreater
+           | TGreaterEqual
+           | TImplies
+           | TLess
+           | TLessEqual
+           | TNot
+           | TUnequal
+
+           | SOF
+           deriving (Eq)
+
+
+data Fixity = Prefix
+            | InfixL
+            | InfixN
+            | InfixR
+            | Postfix
+            deriving (Show)
+
+data Operator = Op Fixity Token
+              | Apply
+              deriving (Show)
+
+operators :: [[Operator]]
```

Comment:
My guess is that this list of lists structure gives the precedence, but it would definitely help to have this invariant documented. (In general, it would help to have more comments :)